File: scripts/validate_structure.py
#!/usr/bin/env python3
# 使い方：$ python scripts/validate_structure.py
import os
import sys

def check_path(path, *, is_dir=False, required=True):
    """
    指定パスが存在するかどうかをチェックする。
    :param path: チェック対象のパス
    :param is_dir: ディレクトリであることを期待する場合 True を指定
    :param required: 存在しない場合にエラーとする場合 True
    :return: 存在していれば True、存在しなければ False
    """
    if is_dir:
        exists = os.path.isdir(path)
    else:
        exists = os.path.isfile(path)
    if not exists and required:
        kind = "directory" if is_dir else "file"
        print(f"Error: Expected {kind} '{path}' not found.")
    return exists

def check_structure():
    errors = False
    base = os.getcwd()  # スクリプト実行時のカレントディレクトリをリポジトリルートとする

    print("Starting directory structure validation...\n")

    # 1. .github 配下のチェック
    github_dir = os.path.join(base, ".github")
    if not os.path.isdir(github_dir):
        print("Error: .github directory not found.")
        errors = True
    else:
        # workflows/validate-structure.yml
        workflows_dir = os.path.join(github_dir, "workflows")
        if not check_path(workflows_dir, is_dir=True):
            errors = True
        else:
            workflow_file = os.path.join(workflows_dir, "validate-structure.yml")
            if not check_path(workflow_file):
                errors = True

        # ISSUE_TEMPLATE/BUG_REPORT.md, FEATURE_REQUEST.md
        issue_template_dir = os.path.join(github_dir, "ISSUE_TEMPLATE")
        if not check_path(issue_template_dir, is_dir=True):
            errors = True
        else:
            for template in ["BUG_REPORT.md", "FEATURE_REQUEST.md"]:
                template_path = os.path.join(issue_template_dir, template)
                if not check_path(template_path):
                    errors = True

    # 2. src/ 配下のチェック
    src_dir = os.path.join(base, "src")
    if not check_path(src_dir, is_dir=True):
        errors = True
    else:
        # [package_name] ディレクトリを特定（src内で __init__.py を持つ最初のディレクトリ）
        package_dir = None
        for entry in os.listdir(src_dir):
            entry_path = os.path.join(src_dir, entry)
            if os.path.isdir(entry_path) and check_path(os.path.join(entry_path, "__init__.py"), required=False):
                package_dir = entry_path
                package_name = entry  # パッケージ名として利用
                break
        if not package_dir:
            print("Error: No Python package directory found in src (directory with __init__.py).")
            errors = True
        else:
            # package内の module1.py の存在確認
            module1 = os.path.join(package_dir, "module1.py")
            if not check_path(module1):
                errors = True

            # subpackage のチェック
            subpackage_dir = os.path.join(package_dir, "subpackage")
            if not check_path(subpackage_dir, is_dir=True):
                errors = True
            else:
                # __init__.py のチェック
                sub_init = os.path.join(subpackage_dir, "__init__.py")
                if not check_path(sub_init):
                    errors = True
                # module_sub.py のチェック（仕様に合わせて必須の場合）
                module_sub = os.path.join(subpackage_dir, "module_sub.py")
                if not check_path(module_sub):
                    errors = True

    # 3. tests/ 配下のチェック
    tests_dir = os.path.join(base, "tests")
    if not check_path(tests_dir, is_dir=True):
        errors = True
    else:
        # 共通テスト設定ファイル
        for test_file in ["conftest.py", "test_module1.py"]:
            if not check_path(os.path.join(tests_dir, test_file)):
                errors = True

        # integration テストディレクトリ
        integration_dir = os.path.join(tests_dir, "integration")
        if not check_path(integration_dir, is_dir=True):
            errors = True
        else:
            for int_file in ["__init__.py", "test_integration.py"]:
                if not check_path(os.path.join(integration_dir, int_file)):
                    errors = True

    # 4. docs/ 配下のチェック
    docs_dir = os.path.join(base, "docs")
    if not check_path(docs_dir, is_dir=True):
        errors = True
    else:
        source_dir = os.path.join(docs_dir, "source")
        build_dir  = os.path.join(docs_dir, "build")
        if not check_path(source_dir, is_dir=True):
            errors = True
        else:
            for doc_file in ["conf.py", "index.rst"]:
                if not check_path(os.path.join(source_dir, doc_file)):
                    errors = True
        if not check_path(build_dir, is_dir=True):
            errors = True

    # 5. scripts/ 配下のチェック
    scripts_dir = os.path.join(base, "scripts")
    if not check_path(scripts_dir, is_dir=True):
        errors = True
    else:
        # このスクリプト自身が scripts/ に存在することをチェック
        current_script = os.path.basename(__file__)
        script_path = os.path.join(scripts_dir, current_script)
        if not os.path.isfile(script_path):
            print(f"Error: Expected file '{current_script}' in scripts directory not found.")
            errors = True

    # 6. ルート直下の主要ファイルチェック
    root_files = [
        ".gitignore",
        "pyproject.toml",
        "requirements.txt",
        "README.md",
        "CODING_GUIDELINES.md",
        "PROJECT_STRUCTURE.md",
        "LICENSE",
        "setup.cfg"
    ]
    for rf in root_files:
        if not check_path(os.path.join(base, rf)):
            errors = True

    print()  # 空行で区切り

    if errors:
        print("Directory structure validation FAILED.")
        sys.exit(1)
    else:
        print("Directory structure validation PASSED.")
        sys.exit(0)

if __name__ == "__main__":
    check_structure()

File: scripts/validate_structure.js
#!/usr/bin/env node
/**
 * validate_structure.js
 *
 * このスクリプトは、プロジェクトのディレクトリ構造が定義どおりになっているかを検証します。
 * 検証対象のディレクトリ・ファイルは、下記の "expectedStructure" オブジェクトに定義されています。
 *
 * 使用方法:
 *   node scripts/validate_structure.js
 */

const fs = require('fs');
const path = require('path');

/**
 * expectedStructure:
 * プロジェクトのルートディレクトリからの相対パスで必要なディレクトリやファイルを定義します。
 *
 * - サブオブジェクトの場合はディレクトリとして扱い、その中のファイル/ディレクトリを検証します。
 * - 値が "file" の場合はファイルとして、"dir" の場合は空のディレクトリとして存在を確認します。
 */
const expectedStructure = {
  ".github": {
    "workflows": {
      "validate-structure.yml": "file"
    },
    "ISSUE_TEMPLATE": {
      "FEATURE_REQUEST.md": "file"
    }
  },
  "src": {
    "main.js": "file",
    "preload.js": "file",
    "renderer": {
      "index.html": "file",
      "renderer.js": "file",
      "style.css": "file"
    }
  },
  "tests": {
    "unit": {
      "test_main.js": "file",
      "test_renderer.js": "file"
    },
    "integration": {
      "test_integration.js": "file",
      "helper.js": "file"
    }
  },
  "dist": "dir",
  "docs": {
    "source": {
      "index.md": "file",
      "config.yml": "file"
    },
    "build": "dir"  // ドキュメント生成物が配置されるディレクトリ（空の場合も許容）
  },
  "scripts": {
    "validate_structure.js": "file",
    "build.js": "file"
  },
  "vendor": {
    "node": "dir"
  },
  "node_modules": "dir",
  ".gitignore": "file",
  "package.json": "file",
  "package-lock.json": "file",
  "README.md": "file",
  "CODING_GUIDELINES.md": "file",
  "PROJECT_STRUCTURE.md": "file",
  "LICENSE": "file",
  "electron-packager-config.json": "file"
};

/**
 * validateStructure
 *
 * @param {string} baseDir - 現在検証中のディレクトリの絶対パス
 * @param {object|string} structureObj - 検証する構造定義。オブジェクトの場合はディレクトリ、文字列の場合は "file" または "dir"
 * @returns {boolean} - 全ての検証が成功した場合 true、失敗があれば false
 */
function validateStructure(baseDir, structureObj) {
  let isValid = true;

  if (typeof structureObj === 'string') {
    // structureObjが単一の "file" または "dir" 指定の場合は、baseDir自体の検証を行う
    if (structureObj === "file") {
      if (!fs.existsSync(baseDir) || !fs.lstatSync(baseDir).isFile()) {
        console.error(`Missing file: ${baseDir}`);
        return false;
      }
    } else if (structureObj === "dir") {
      if (!fs.existsSync(baseDir) || !fs.lstatSync(baseDir).isDirectory()) {
        console.error(`Missing directory: ${baseDir}`);
        return false;
      }
    }
    return true;
  }

  // 構造がオブジェクトの場合、キーを順次検証
  for (const key in structureObj) {
    const expected = structureObj[key];
    const fullPath = path.join(baseDir, key);

    if (typeof expected === "string") {
      // 単一のファイルまたはディレクトリの存在確認
      if (expected === "file") {
        if (!fs.existsSync(fullPath) || !fs.lstatSync(fullPath).isFile()) {
          console.error(`Missing file: ${fullPath}`);
          isValid = false;
        }
      } else if (expected === "dir") {
        if (!fs.existsSync(fullPath) || !fs.lstatSync(fullPath).isDirectory()) {
          console.error(`Missing directory: ${fullPath}`);
          isValid = false;
        }
      }
    } else if (typeof expected === "object") {
      // オブジェクトの場合はディレクトリとして存在するか確認
      if (!fs.existsSync(fullPath) || !fs.lstatSync(fullPath).isDirectory()) {
        console.error(`Missing directory: ${fullPath}`);
        isValid = false;
      } else {
        // 再帰的に内部の構造を検証
        const subValid = validateStructure(fullPath, expected);
        if (!subValid) {
          isValid = false;
        }
      }
    }
  }
  return isValid;
}

/**
 * main
 *
 * リポジトリのルートディレクトリから検証を実施します。
 */
function main() {
  // スクリプトが配置されているディレクトリから見たリポジトリのルート（例: scripts/から1階層上）
  const repoRoot = path.resolve(__dirname, '..');
  const structureValid = validateStructure(repoRoot, expectedStructure);

  if (!structureValid) {
    console.error("Project structure validation failed.");
    process.exit(1);
  } else {
    console.log("Project structure validated successfully.");
    process.exit(0);
  }
}

main();
